### PR TITLE
Web scraper service

### PR DESCRIPTION
**Pull request by t-kanin**

### CODE REVIEW

File: app/services/scraper_service.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+require 'nokogiri'
+
+class ScraperService < ApplicationService
+  attr_reader :keywords
+
+  def initialize(keywords)
+    @keywords = keywords
+  end
+
+  def call
```
Comment (WatcharitNew) [inline app/services/scraper_service.rb]: You can do something like this
```suggestion
class ScraperService < ApplicationService
  include Singleton
  
  def call(keyword)
```
and when you call `ScraperService` you use this instead `ScraperService.instance.call(keyword)`

File: Gemfile [hunk 2]
```diff
@@ -31,7 +31,10 @@ gem 'semantic-ui-sass'
 # Reduces boot times through caching; required in config/boot.rb
 gem 'bootsnap', '>= 1.4.4', require: false
 gem 'devise', '~> 4.8'
+gem 'nokogiri', '~> 1.12', '>= 1.12.4'
 gem 'selenium-webdriver', '~> 3.142', '>= 3.142.7'
+gem 'sidekiq', '~> 6.2', '>= 6.2.2'
+gem 'sidekiq-cron', '~> 1.2'
```
Comment (aunsira) [inline Gemfile]: Remove if not used.

File: spec/workers/scraper_worker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+RSpec.describe ScraperWorker, type: :worker do
```
Comment (aunsira) [inline spec/workers/scraper_worker_spec.rb:2]: Please work on unit test as well (could be next PR or something)

File: config/sidekiq.yml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+:pidfile: ./tmp/pids/sidekiq.pid
+:logfile: ./log/sidekiq.log
+:concurrency: 10
+
```
Comment (aunsira) [inline config/sidekiq.yml]: Extra empty line.

File: app/services/scraper_service.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  def search_result(doc)
+    res = []
+    elements = doc.css('div.g')
+    elements.pop
+    elements.each do |e|
+      h = { title: e.css('h3').text, link: e.css('a @href').first.text }
+      res << h
+    end
+    res
```
Comment (aunsira) [inline app/services/scraper_service.rb]: Would comeback to this later.

File: app/services/scraper_service.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+require 'selenium-webdriver'
+require 'nokogiri'
```
Comment (aunsira) [inline app/services/scraper_service.rb]: Might not need to import this?

File: app/services/scraper_service.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  def search(doc, div, title, link)
+    res = []
+    elements = doc.css(div)
+    elements.each do |e|
+      res << { title: e.css(title).text, link: e.css(link).first.text } unless e.css(title).blank?
+    end
+
+    res
+  end
```
Comment (aunsira) [inline app/services/scraper_service.rb]: What about this refactoring
```diff
diff --git a/app/services/scraper_service.rb b/app/services/scraper_service.rb
index e92bc8a..524140f 100644
--- a/app/services/scraper_service.rb
+++ b/app/services/scraper_service.rb
@@ -24,22 +24,20 @@ class ScraperService < ApplicationService
 
   def result(driver, keyword)
     driver.find_element(name: 'q').send_keys keyword, :return
-    doc = Nokogiri::HTML(driver.page_source)
+    @doc = Nokogiri::HTML(driver.page_source)
     {
       page_result: page_result(doc),
-      search_result: search(doc, 'div.g', 'h3', 'a @href'),
-      ad_result: search(doc, 'div#tads a', 'span', '@href')
+      search_result: fetch_non_adword_links
+      ad_result: fetch_top_position_adwords_links
     }
   end
 
-  def search(doc, div, title, link)
-    res = []
-    elements = doc.css(div)
-    elements.each do |e|
-      res << { title: e.css(title).text, link: e.css(link).first.text } unless e.css(title).blank?
-    end
+  def fetch_top_position_adwords_links
+    @dic.css('div#tads a').map { |link| { title: link.css('span').text, link: link.css('@href').first.text } }.compact
+  end
 
-    res
+  def fetch_non_adword_links
+    @dic.css('div.g').map { |link| { title: link.css('h3').text, link: link.css('a @href').first.text } }.compact
   end
 
   def page_result(doc)
```

File: app/services/scraper_service.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+class ScraperService < ApplicationService
+  include Singleton
+
+  def call(keywords)
+    key_res = []
+    driver = init_driver
+    keywords.each do |keyword|
+      res = result(driver, keyword)
+      driver.find_element(name: 'q').clear
+      key_res << res
```
Comment (aunsira) [inline app/services/scraper_service.rb]: You may want to get help from instance variable.
```diff
diff --git a/app/services/scraper_service.rb b/app/services/scraper_service.rb
index e92bc8a..c6467d3 100644
--- a/app/services/scraper_service.rb
+++ b/app/services/scraper_service.rb
@@ -4,27 +4,26 @@ class ScraperService < ApplicationService
   include Singleton
 
   def call(keywords)
-    key_res = []
-    driver = init_driver
-    keywords.each do |keyword|
-      res = result(driver, keyword)
-      driver.find_element(name: 'q').clear
-      key_res << res
+    init_driver
+    result = keywords.each_with_object([]) do |keyword, accum|
+      res = result keyword
+      @driver.find_element(name: 'q').clear
+      accum << res
     end
 
-    driver.quit
-    key_res
+    @driver.quit
+    result
   end
 
   def init_driver
-    driver = Selenium::WebDriver.for :chrome
-    driver.get 'http://www.google.com/'
-    driver
+    @driver = Selenium::WebDriver.for :chrome
+    @driver.get 'http://www.google.com/'
+    @driver
   end
 
   def result(driver, keyword)
-    driver.find_element(name: 'q').send_keys keyword, :return
-    doc = Nokogiri::HTML(driver.page_source)
+    @driver.find_element(name: 'q').send_keys keyword, :return
+    doc = Nokogiri::HTML(@driver.page_source)
     {
       page_result: page_result(doc),
       search_result: search(doc, 'div.g', 'h3', 'a @href'),
```

File: app/services/scraper_service.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        .reject { |h| h[:title].empty? }
+  end
+
+  def fetch_non_adwords_links
+    @dic.css('div.g').map { |link| { title: link.css('h3').text, link: link.css('a @href').first.text } }
+        .reject { |h| h[:title].empty? }
+  end
+
+  def page_result
+    @dic.at_css('div#result-stats').text
```
Comment (aunsira) [inline app/services/scraper_service.rb:47]: 🙈 it was my typo

Comment (t-kanin) [reply app/services/scraper_service.rb:47]: thought it was naming convention, so changed it accordingly

**Review by aunsira — APPROVED** [review]

**Pull request merged**